src/compiler/checker.ts

```diff
@@ -6011,7 +6018,70 @@ namespace ts {
                             case SyntaxKind.IfStatement:
                                 // In a branch of an if statement, narrow based on controlling expression
                                 if (child !== (<IfStatement>node).expression) {
+
+                                    // Union types that got narrowed to one type is no longer union types. We want to return a
+                                    // void type in case if the type returned by one if-statement clause is the same as the
+                                    // narrowed type, given that the original type was a union type.
+                                    //
+                                    // Example:
+                                    //
+                                    //     let x: A | B | C;
+                                    //
+                                    //     if (isA(x)) { // isA(...) returns A and the narrowing type has the value of A so narrow to void type.
+                                    //
+                                    //     }
+                                    //     else if (isB(x)) { //  narrow to A
+                                    //
+                                    //     }
+                                    //     else if (isC(x)) { // narrow to A | B
+                                    //
+                                    //     }
+                                    //     else { // type begins with A | B | C
+                                    //
+                                    //          x // is void
+                                    //     }
+                                    //
+                                    if (!(type.flags & TypeFlags.Union)) {
+                                        // Get the original type again because we don't want to store it on a variable higher
+                                        // up to minimize memory allocation.
+                                        let originalType = getTypeOfSymbol(symbol);
+
+                                        if (originalType.flags & TypeFlags.Union &&
+                                            // Use the original type to check the narrowed type instead of the ongoing narrowing type.
+                                            // This is to deal with cases like below:
+                                            //
+                                            //     if (isB(x)) // narrow to A.

```

Comment:
The comment wasn't so clear but it remove `B` from a set that only contains `A`. So the returned narrowed type is `A`.